File: 0053-maximum-subarray/0053-maximum-subarray.py
class Solution:
    def maxSubArray(self, nums: List[int]) -> int:
        max_sum = nums[0]
        current_sum = 0

        for num in nums:
            if current_sum < 0:  # Do not carry -ve sums forward
                current_sum = 0  # Reset the front end of the sliding window

            current_sum += num
            max_sum = max(max_sum, current_sum)

        return max_sum

File: 0053-maximum-subarray/0053-maximum-subarray.cpp
class Solution {
public:
    int maxSubArray(vector<int>& nums) {
        int n = nums.size();
        
        int max_sum = INT_MIN, current_sum = 0;
        for (int i = 0; i < n; i++) {
            current_sum += nums[i];
            
            max_sum = max(max_sum, current_sum);
            current_sum = max(current_sum, 0); // If overall sum is -ve, then reset to 0
        }
        
        return max_sum;
    }
};